nixos-rebuild-ng: implement `--target-host`

thiagokokada opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

This PR ended up bigger than I expected because it needed lots of changes to support remote builders. Why this took so much code? Well:
- Need to differentiate between the multiple possible Nix flags, e.g.: `--use-substitutes` is only for `nix-copy-closure` while `--include` is only for `nix-build`
- Need an abstraction to run command remotely, so now we have `process.run_wrapper`
- We need some things done completely different remotely, e.g.: get hostname

However since everything is here now, it should be easy to add `--build-host` next. But `--build-host` is sufficiently different that I want to do this in a separate PR, e.g.: we can't use `nix-build` and need to use `nix-copy-closure` and `nix-store -r` to  realise the store path.

~~This is basically a bug to bug implementation of the original. There is an issue with `--use-remote-sudo` asking for password twice that is also reproducible in the original. It can be fixed, however I don't think it is worth it considering that we eventually we will have `activate` script once it gets back into the tree, and once we have `activate` this issue goes away.~~

I changed the implementation of remote to not depend in `ssh -t` (pseudo TTY) anymore. This is a headache since it changes lots of behaviors inside SSH (for example, it redirects all the output to stdout), and it is also unreliable.

However, `sudo` depends in the TTY to ask for password, so now we have a flag `--ask-sudo-password` that will ask the password before running the tasks remotely. For local this doesn't change anything.

I am not sure if there is some concerns about security here: we are storing the password in memory and there is no good way in Python to remove it (like `memset` that I assume `sudo` uses here), so technically it will stay there until some program overwrites it. I don't think it is a huge concern though, people store secrets in environment variables all the time and this is probably easier to dump than memory. Also, if someone wants to grab your sudo password and have access to your computer to do so, they could run a keylogger and it would be much easier than dumping your memory. But if you are a reviewer and have some concerns here, please say so. 

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

thiagokokada committed:
nixos-rebuild-ng: explicitly parse Nix flags

thiagokokada committed:
nixos-rebuild-ng: add --sudo/--use-remote-sudo flags

thiagokokada committed:
nixos-rebuild-ng: create instance for dataclass from Self

thiagokokada committed:
nixos-rebuild-ng: run -> run_wrapper, handle encode errors and add extra_env

thiagokokada committed:
nixos-rebuild-ng: add TTY allocation in SSH

thiagokokada committed:
nixos-rebuild-ng: parse NIX_SSHOPTS instead of SSH_OPTS env var

thiagokokada committed:
nixos-rebuild-ng: implement `--target-host`

thiagokokada committed:
nixos-rebuild-ng: add pythonpath to pytest config

thiagokokada committed:
nixos-rebuild-ng: error when `--rollback` is called with incompatible action

thiagokokada committed:
nixos-rebuild-ng: add allow_tty parameter to process.run_wrapper

thiagokokada committed:
nixos-rebuild-ng: get remote hostname

thiagokokada committed:
nixos-rebuild-ng: move models.Ssh to process.Remote

thiagokokada committed:
nixos-rebuild-ng: remove support for env in process.run_wrapper

thiagokokada committed:
nixos-rebuild-ng: cleanup SSH ControlMaster at exit

thiagokokada committed:
nixos-rebuild-ng: implement `--target-host` for `--rollback`

thiagokokada commented:
CC @Scrumplex for review.

thiagokokada commented:
@ofborg build nixos-rebuild-ng.tests

lucasew commented on pkgs/by-name/ni/nixos-rebuild-ng/src/nixos_rebuild/utils.py:
```diff
@@ -1,18 +1,18 @@
         match value:
-            case None | False:
+            case None | False | 0 | []:
                 pass
```
continue

thiagokokada replied:
Done.

lucasew commented on pkgs/by-name/ni/nixos-rebuild-ng/src/tests/test_models.py:
```diff
@@ -98,3 +111,37 @@ def test_profile_from_name(mock_mkdir: Any) -> None:
+def test_remote_from_name(monkeypatch: Any, tmpdir: Path) -> None:
+    monkeypatch.setenv("NIX_SSHOPTS", "")
+    assert m.Remote.from_arg("user@localhost", None, tmpdir) == m.Remote(
+        "user@localhost",
```
use keyword arguments to make it explicit

thiagokokada committed:
nixos-rebuild-ng: move test to the correct file

thiagokokada committed:
nixos-rebuild-ng: do not use TTY for `--target-host`

Instead this commit introduces the `--ask-sudo-password` that stores the
password in memory and injects it via `stdin` if the user wants.

thiagokokada committed:
nixos-rebuild-ng: remove explicit check for git and instead check exception

thiagokokada committed:
nixos-rebuild-ng: split parser_args in get_parser

thiagokokada committed:
nixos-rebuild-ng: import nix module instead of each individual function

thiagokokada committed:
nixos-rebuild-ng: set process.run_wrapper check=True by default

thiagokokada commented:
@ofborg build nixos-rebuild-ng.tests

thiagokokada commented:
@ofborg build nixos-rebuild-ng nixos-rebuild-ng.tests

thiagokokada committed:
nixos-rebuild-ng: validate NIX_SSHOPTS

Scrumplex commented on pkgs/by-name/ni/nixos-rebuild-ng/src/nixos_rebuild/__init__.py:
```diff
@@ -112,7 +112,7 @@ def parser_warn(msg):
     # TODO: use deprecated=True in Python >=3.13
     if args.no_ssh_tty:
-        parser_warn("--no-ssh-tty deprecated, SSH's pseudo-TTY is never used anymore")
+        parser_warn("--no-ssh-tty deprecated, SSH's TTY is never used anymore")
```
Isn't it technically still used for the sudo prompt? Maybe reword this:

`--no-ssh-tty deprecated, a TTY is only allocated if it's explicitly needed`

thiagokokada replied:
No, SSH's TTY (the `-t` flag) is never used anymore, this is why we have `--ask-sudo-password` (that asks the password using Python's `getpass` instead and pass the password via stdin using `--stdin` flag in `sudo`).

The reason for it is because SSH's TTY mess up with lots of the expectations you have with SSH, e.g.: it combines stdout and stderr in one stream (stdout), so if you expect a command to print something to stderr for example it will be wrong.

Of course this can be workaround, but it is better to simple not do it. The bonus is that using something like `--ask-sudo-password` means we avoid the previous issue where the `sudo` password were asked twice because each SSH connection created a new TTY and by default the SSH cache is associated to each TTY (it can be changed, but it is rarely done so). Also this is the way other tools do it, for example Ansible has a `--ask-become-pass` that is similar behavior of `--ask-sudo-password`.

Scrumplex replied:
Oh, I see. I misread the code earlier.

thiagokokada committed:
nixos-rebuild-ng: use argparse groups to group nix flags

thiagokokada commented:
Pushed a commit to simplify how the Nix arguments are declared. Instead of needing to group them twice, now we are using the Argparse groups to group the arguments. This avoid declaration of the groups twice.

thiagokokada commented:
Will stop forcing push for now, I am doing that bad habit of just rewriting code since refactoring is fun 😅.

CC @Scrumplex for review.

lucasew commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 359097`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>nixos-rebuild-ng</li>
    <li>nixos-rebuild-ng.dist</li>
  </ul>
</details>

lucasew commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 359097`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>nixos-rebuild-ng</li>
    <li>nixos-rebuild-ng.dist</li>
  </ul>
</details>

Scrumplex reviewed (commented):
Changes LGTM. I will test this and report back in a moment
